qcodes_contrib_drivers/drivers/Ametek/SR_7270.py

```diff
@@ -0,0 +1,296 @@
+                                        200e-3: 25,
+                                        500e-3: 26,
+                                        1: 27},
+                        docstring=("Set measurement input sensitivity; "
+                                   "only settable."))
+
+        self.add_parameter(name='timeconstant',
+                        label='Time constant',
+                        unit='s',
+                        get_cmd='TC',
+                        set_cmd='TC {}',
+                        val_mapping={   10e-6: 0,
+                                        20e-6: 1,
+                                        50e-6: 2,
+                                        100e-6: 3,
+                                        200e-6: 4,
+                                        500e-6: 5,
+                                        1e-3: 6,
+                                        2e-3: 7,
+                                        5e-3: 8,
+                                        10e-3: 9,
+                                        20e-3: 10,
+                                        50e-3: 11,
+                                        100e-3: 12,
+                                        200e-3: 13,
+                                        500e-3: 14,
+                                        1: 15,
+                                        2: 16,
+                                        5: 17,
+                                        10: 18,
+                                        20: 19,
+                                        50: 20,
+                                        100: 21,
+                                        200: 22,
+                                        500: 23,
+                                        1e+3: 24,
+                                        2e+3: 25,
+                                        5e+3: 26,
+                                        10e+3: 27,
+                                        20e+3: 28,
+                                        50e+3: 29,
+                                        100e+3: 30},
+                        docstring=("Set measurement time constant; "
+                                   "only settable."))
+
+    def ask_raw(self, cmd:str) -> str:
+        """Reimplementaion of ask function to handle lockin echo.
+
+        Args:
+            cmd: Command to be sent (asked) to lockin.
+
+        Raises:
+            Runtimeerror: If the response does not end with the expected terminators '\n\x00' or '\x00'
+
+        Returns:
+            str: Return string from lockin with terminator character stripped of.
+        """
+        with DelayedKeyboardInterrupt():
+            response = self.visa_handle.query(cmd)
+            if response.endswith('\x00'):
+                resp = response[:-1]
+                if resp.endswith('\n'):
+                    return resp[:-1]
+                else:
+                    return resp
+            else:
+                raise RuntimeError(response)
```

Comment:
i'd recommend to reimplement this with using ``super().ask_raw(...)`` and then add additional processing if needed. Moreover, i don't think this additional processing is needed - the pyVISA can do this for you if you set the terminator to be '\n\x00' - have you tried that?

Reply:
@astafan8 I have found some time to look into this. We do the processing of the response due to the fact that the device not always responds with the same terminator. Read commands are terminated with "\n\x00" and write commands are acknowledged with a return of only "\x00".
So you say that a single terminator of "\n\x00" would still work?

Concerning the reimplementation using `super().ask_raw(...)`: this means you would replace `response = self.visa_handle.query(cmd)` with `response = super().ask_raw(cmd)`?

Reply:
> oncerning the reimplementation using super().ask_raw(...): this means you would replace response = self.visa_handle.query(cmd) with response = super().ask_raw(cmd)?

yes :)

> Read commands are terminated with "\n\x00" and write commands are acknowledged with a return of only "\x00".
So you say that a single terminator of "\n\x00" would still work?

oh, i see. No, single terminator is unlikely to work here. What you can do is to set read and write terminators individually on the visa_handle level like here: https://github.com/QCoDeS/Qcodes/blob/c596700363ff23967c10ea49d3161f614ca119b7/src/qcodes/instrument/visa.py#L234C1-L235C59 . You can implement this in the following way: in the `__init__` of the driver, when you call `super().__init__(....)`, pass `terminator=None` as the argument so that the qcodes VisaInstrument does not set the terminator at all, and then after that set the write and read terminators via `visa_handle`. It would also be a good practice to add read_terminator and write_terminator as keyword arguments to the driver that you're writing so that (just in case) the terminators can be changed by the user of the driver.

Reply:
Ok, then if I understood correctly, we would not need at all the redefinition of `ask_raw `and `write_raw `functions since pyvisa would take care of the different terminators.
I will have to check this in the lab with an actual device.

Reply:
I did not manage to solve this issue with the terminator. The hack checking for the terminators in the `ask_raw` function works fine and I will not dig deeper to solve this. I hope otherwise the driver is fine.

Reply:
ok